File: diceware-java/src/main/java/diceware.java
/**
 * Created by Jacob on 1/3/2016.
 */

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.Type;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.SecureRandom;
import java.util.Map;

public class diceware {

    public static void main(String[] args) {

        Map<String, String> dicewareMap = getDicewareMap();
        SecureRandom sr = getSecureRandom();

        int n = 5; // default
        try {
            n = Integer.parseInt(args[0]);
        } catch (Exception ignored) {}

        for (int i = 0; i < n; i++) {
            String rollKey = fiveRolls(sr);
            System.out.println(dicewareMap.get(rollKey));
        }
    }

    // Return a string represtenting a roll of five dice, such as "12346"
    private static String fiveRolls(SecureRandom sr) {
        StringBuilder sb = new StringBuilder("");
        for (int i = 0; i < 5; i++) {
            int roll = sr.nextInt(6) + 1;
            sb.append(roll);
        }
        return sb.toString();
    }

    // Get dicewareMap from resource (jar).
    private static Map<String,String> getDicewareMap() {
        Gson gson = new Gson();
        InputStream IS = diceware.class.getResourceAsStream("diceware-map.json");
        JsonReader reader;
        try {
            reader = new JsonReader(new InputStreamReader(IS, "ASCII"));
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
        Type dicewareWordPairType = new TypeToken<Map<String, String>>() {}.getType();
        Map<String, String> dicewareMap = gson.fromJson(reader, dicewareWordPairType);
        return dicewareMap;
    }

    // Perpare to choose random numbers.
    // https://www.cigital.com/blog/proper-use-of-javas-securerandom/
    private static SecureRandom getSecureRandom() {
        SecureRandom sr;
        try {
            try {
                sr = SecureRandom.getInstance("SHA1PRNG", "SUN");
            } catch (NoSuchProviderException e) {
                sr = SecureRandom.getInstance("SHA1PRNG");
            }
        } catch (NoSuchAlgorithmException e1) {
            sr = new SecureRandom();
        }
        sr.nextBytes(new byte[8]);
        return sr;
    }


}

File: diceware-kotlin/src/main/kotlin/demo/diceware.kt
/**
 * Created by Jacob on 1/8/2016.
 */

package demo

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import java.io.InputStreamReader
import java.security.NoSuchAlgorithmException
import java.security.NoSuchProviderException
import java.security.SecureRandom

fun main(args: Array<String>) {
    val dicewareMap = DicewareMap.getDicewareMap()
    val sr = getSecureRandom()

    var n = 5  // default
    try {
        n = Integer.parseInt(args[0])
    } catch(ignored: Exception) {}

    for (i in 1..n) {
        var rollKey = fiveRolls(sr);
        println(dicewareMap.get(rollKey));
    }
}

fun fiveRolls(sr: SecureRandom): String {
    val sb: StringBuilder = StringBuilder("");
    for (i in 1..5) {
        var roll = sr.nextInt(6) + 1;
        sb.append(roll);
    }
    return sb.toString();
}

object DicewareMap {
    fun getDicewareMap(): Map<String, String> {
        val gson = Gson()
        val IS = DicewareMap::class.java.getResourceAsStream("/diceware-map.json")
        val reader = JsonReader(InputStreamReader(IS, "ASCII"))
        val dicewareMap: Map<String, String> = gson.fromJson(reader)
        return dicewareMap
    }
}

// This extension function for the Gson instance looks at the desired
// return type, and fills in the appropriate TypeToken
// for the "regular" java call to #fromJson(reader,type)
// http://stackoverflow.com/a/33420043/3599738
inline fun <reified T> Gson.fromJson(json: JsonReader): T {
    return this.fromJson<T>(json, object : TypeToken<T>() {}.type)
}

fun getSecureRandom(): SecureRandom {
    var sr: SecureRandom;
    try {
        try {
            sr = SecureRandom.getInstance("SHA1PRNG", "SUN")
        } catch(e: NoSuchProviderException) {
            sr = SecureRandom.getInstance("SHA1PRNG")
        }
    } catch(e: NoSuchAlgorithmException) {
        sr = SecureRandom()
    }
    sr.nextBytes(ByteArray(8))
    return sr
}